Principal Token Standard (#5089)
ethereum/EIPs: EIPS/eip-5089.md

---
eip: 5089

> title: Principal Token Standard
> description: A standard for principal tokens (zero-coupon tokens) that are redeemable for a single underlying ERC-20 token at a future timestamp.

Reviewer: We discourage "standard" in the title and description of EIPs:

```suggestion
title: Principal Token
description: Principal tokens (zero-coupon tokens) are redeemable for a single underlying ERC-20 token at a future timestamp.
```

Author: Fixed

author: Julian Traversa (@JTraversa), Robert Robbins (@robrobbins)

> discussions-to: https://ethereum-magicians.org/t/eip-xxxx-principal-token-standard/1

Reviewer: This link doesn't exist.

Author: Fixed, thank you.

status: Draft
type: Standards Track
category: ERC
created: 2022-05-01
requires: 20, 2612
---

## Abstract

> The following standard allows for the implementation of a standard API for principal tokens --
> representing ownership of an underlying [ERC-20](./eip-20.md) token at a future timestamp.
> This standard is an extension on the ERC-20 token that provides basic functionality for depositing
> and withdrawing tokens and reading balances and the ERC-2612 standard that provides
> EIP-712 signature based approvals.

Reviewer: Some general copyediting, plus adding links to mentioned EIPs:

```suggestion
The following standard defines an API for principal tokens (also known as zero-coupon tokens) that represent ownership of an underlying [ERC-20](./eip-20.md) token at a future timestamp. This standard is an extension on the ERC-20 token that provides basic functionality for depositing/withdrawing tokens, and reading balances; and the [ERC-2612](./eip-2612.md) standard that provides [EIP-712](./eip-712.md) signature based approvals.
```

Author: Fixed

## Motivation

Principal tokens have an increasing lack of standardization which has led to a difficult to navigate development space and diverse implementation
schemes.

The primary examples include yield tokenization platforms which strip future yield leaving a principal
token behind, as well as fixed-rate money-markets which utilize principal tokens as a medium
to lend/borrow.

This inconsistency in implementation makes integration difficult at the application layer as well as
wallet layer which are key catalysts for the space's growth. 
Developers are currently expected to implement individual adapters for each principal token, as well as adapters for
their pool contracts, and many times adapters for their custodial contracts as well, wasting significant developer resources. 

## Specification

> All ERC-5070 Principal Tokens MUST implement ERC-20 to represent ownership of future underlying redemption (PTs).

Reviewer: Wrong EIP number, and incorrect abbreviation placement?

```suggestion
Each ERC-5089 Principal Token (PT) MUST implement ERC-20 to represent ownership of future underlying redemption.
```

Author: Fixed

If a PT is to be non-transferrable, it MAY revert on calls to `transfer` or `transferFrom`.
The ERC-20 operations `balanceOf`, `transfer`, `totalSupply`, etc. operate on the Principal Token balance
which represents a claim to ownership of 1:1 amounts of underlying upon maturity.

All ERC-5070 Principal Tokens MUST implement ERC-20's optional metadata extensions.
The `name` and `symbol` functions SHOULD reflect the underlying token's `name` and `symbol` in some way, as well as the origination protocol, 
and in the case of yield tokenization protocols, the origination money-market.

ERC-5070 tokenized Vaults MAY implement [EIP-2612](./eip-2612.md) to improve the UX of approving PTs on various integrations.

### Definitions:

- underlying: The token that Principal Tokens are redeemable for at maturity.
  Has units defined by the corresponding ERC-20 contract.
- maturity: The timestamp (unix) at which a Principal Token matures. Principal Tokens become redeemable for underlying at or after this timestamp.
- exchangeRate: A variable used to modulate the redeemable ratio of underlying <-> Principal Token.

> - fee: An amount of underlying or Principal Token charged to the user by the Principal Token. Fees can exists on redemption or post-maturity yield. _Alberto: I'm not sure we need this in the ERC._

Reviewer 2: What were the arguments for it being in 4626?

I didn't think it deserved to be there either but it seemed like folks wanted to futureproof the standard. Contextually it makes even less sense here but just curious.

Author: In 4626 it was assumed that some implementations would definitely include fees on deposits or withdrawals. I agree that maybe fees don't make much sense for principal tokens.

> - fee: An amount of underlying or Principal Token charged to the user by the Principal Token. Fees can exists on redemption or post-maturity yield. _Alberto: I'm not sure we need this in the ERC._

Reviewer: ```suggestion
- fee: An amount of underlying or Principal Token charged to the user by the Principal Token. Fees can exists on redemption or post-maturity yield.
```

### Methods

#### underlying

The address of the underlying token used by the Principal Token for accounting, depositing, and withdrawing.

MUST be an ERC-20 token contract.

MUST _NOT_ revert.

```yaml
- name: underlying
  type: function
  stateMutability: view

  inputs: []

  outputs:
    - name: underlyingAddress
      type: address
```

#### maturity

The unix timestamp (uint256) after which Principal Tokens can be redeemed for their underlying deposit according to `exchangeRate`.

MUST _NOT_ revert.

```yaml
- name: maturity
  type: function
  stateMutability: view

  inputs: []

  outputs:
    - name: timestamp
      type: uint256
```

#### maturityRate

The exchangeRate during the block that a first user first calls `redeem` or `mature` after maturity.

> OPTIONAL - This method can be used to provide needed data for the calculation of interest post-maturity, but interfaces and other contracts MUST NOT expect this method to be present. _Alberto: Optionality in ERCs is better avoided. Not everyone will implement it, therefore integrators must assume it might not be there, and leads to a proliferation of `safe*` methods._

Reviewer: ```suggestion
OPTIONAL - This method can be used to provide needed data for the calculation of interest post-maturity, but interfaces and other contracts MUST NOT expect this method to be present.
```

MUST _NOT_ revert.

```yaml
- name: maturityRate
  type: function
  stateMutability: view

  inputs: []

  outputs:
    - name: rate
      type: uint256
```

#### exchangeRate

The current exchangeRate of the underlying token <-> money market deposit post-maturity. Returns 0 before maturity.

> _Alberto: I wonder if we could make the ERC a bit more compact, and have only `exchangeRate` (renamed to `rate`). Before maturity, it returns the rate that will work at maturity, after maturity, it can return the same, or change._

Reviewer: ```suggestion
```

> OPTIONAL - This method can be used to provide needed data for the calculation of interest post-maturity, but interfaces and other contracts MUST NOT expect this method to be present. _Alberto: Optionality in ERCs is better avoided. Not everyone will implement it, therefore integrators must assume it might not be there, and leads to a proliferation of `safe*` methods._

Reviewer: ```suggestion
OPTIONAL - This method can be used to provide needed data for the calculation of interest post-maturity, but interfaces and other contracts MUST NOT expect this method to be present.
```

MUST _NOT_ revert.

```yaml
- name: exchangeRate
  type: function
  stateMutability: view

  inputs: []

  outputs:
    - name: rate
      type: uint256
```

#### mint

> _Alberto: do we need a `mint` method? I reckon it might be nice for integrators to have a standard function to obtain the PTs, but so far we have been fine without one for ERC20s._

Reviewer 2: @alcueca Yeah I am unsure if a `mint` is necessary.

For yield tokenization protocols, `mint` is generally on a central router contract, given when minting you are also minting Yield Tokens as well, while in fixed-yield money-markets, a central contract handles usually bookkeeping and thus minting as well.

That said, I noticed its implementation in a couple codebases (including Yield's) so included it since redundant minting capabilities through the PT could be marginally useful for Yield Tokenization protocols and it seemed you guys had already included it. Was there any specific rationale towards having `mint` in the fyTokens?

Author: We included it in the fyToken as a workaround for some issues with our liquidity rolling contracts. Basically they mint fyToken out of udnerlying, and it was a lot simpler to do it that way, while still correct. It is pretty much an edge case that shouldn't be representative of anything intrisic to fixed yield tokens.

> _Alberto: do we need a `mint` method? I reckon it might be nice for integrators to have a standard function to obtain the PTs, but so far we have been fine without one for ERC20s._

Reviewer: ```suggestion
```

Please move discussion to the ethereum magicians thread.

> Mints exactly `amount` of Principal Tokens to `to` by depositing that exact same `amount` of underlying tokens. In yield tokenization protocols, also mints `amount` of Yield Tokens in addition to Principal Tokens.

Reviewer 3: This is not entirely true for yield tokenization protocols. In case there is some yield already in the pool, Yield token can have value `>0` so if you deposit `amount` of underlying you can get less PT + YT.

While custodial minting of Principal Tokens through the Principal Token contract is extremely useful for integrators, some protocols may find giving the Principal Token itself custody breaks their backwards compatability. 

MUST emit the `Mint` event.

MUST support ERC-20 `approve` / `transferFrom` on `asset` as a mint flow.
MAY support an additional flow in which the underlying tokens are owned by the Principal Token contract before the `mint` execution, and are accounted for during `mint`.

MUST revert if all of `amount` cannot be minted (due to deposit limit being reached, slippage, the user not approving enough underlying tokens to the PT contract, etc).

Note that most implementations will require pre-approval of the Principal Token contract with the Principal Tokens `underlying` token.

```yaml
- name: mint
  type: function
  stateMutability: nonpayable

  inputs:
    - name: amount
      type: uint256
    - name: to
      type: address

  outputs:
    - name: success
      type: bool
```

#### redeem

After maturity, burns exactly `amount` of Principal Tokens from `from` and sends the _same_ `amount` of underlying tokens to `to`.

> _Alberto: It sends underlying tokens according to `exchangeRate`._

Reviewer 2: >  Optionality is evil.

I didn't want to force *every* protocol's implementation to distribute yield post maturity, would this mean that these protocols are just expected to set an `exchangeRate` of 1:1 at maturity?

Seems like an ok option, though it would mean compliance with the standard just for compliance's sake.

Anywho, assuming thats the direction the EIP goes, then yeah 100% on board with injecting the language everywhere.

> After maturity, burns exactly `amount` of Principal Tokens from `from` and sends the _same_ `amount` of underlying tokens to `to`.

-> 

After maturity, burns exactly `amount` of Principal Tokens from `from` and sends underlying tokens according of underlying tokens to the current exchangeRate.

> _Alberto: It sends underlying tokens according to `exchangeRate`._

Reviewer: ```suggestion
```

Author: Done

OPTIONAL - This method can be used to improve the usability and integration difficulty of Principal Tokens, but interfaces and other contracts MUST NOT expect fund custody to be present.
While custodial redemption of Principal Tokens through the Principal Token contract is extremely useful for integrators, some protocols may find giving the Principal Token itself custody breaks their backwards compatability. 

> _Alberto: Optionality is evil._

Reviewer: ```suggestion
```

MUST emit the `Redeem` event.

MUST support a redeem flow where the Principal Tokens are burned from `owner` directly where `owner` is `msg.sender` or `msg.sender` has ERC-20 approval over the shares of `owner`.
MAY support an additional flow in which the shares are transferred to the Principal Token contract before the `redeem` execution, and are accounted for during `redeem`.

MUST revert if all of `amount` cannot be redeemed (due to withdrawal limit being reached, slippage, the owner not having enough Principal Tokens, etc).

Note that some implementations will require pre-requesting to the Principal Token before a withdrawal may be performed. Those methods should be performed separately.

```yaml
- name: redeem
  type: function
  stateMutability: nonpayable

  inputs:
    - name: amount
      type: uint256
    - name: to
      type: address
    - name: from
      type: address

  outputs:
    - name: assets
      type: uint256
```

#### mature

> _Alberto: We surely don't need this method in the standard. It is up to each PT creator to decide how they mature it._

Reviewer 2: `mature` was included to establish the workflow of an `exchangeRate` at maturity. 

I think if we instead move towards an extension of 4626 or similar, the `convertTo` methods would mean explicitly setting an `exchangeRate` at maturity isnt necessary as it becomes an implementation detail, and having `mature` in the standard would be unnecessary.

Author: Gove this way with the latest revision. In general, we should not include anything forcing a particular implementation.

> _Alberto: We surely don't need this method in the standard. It is up to each PT creator to decide how they mature it._

Reviewer: ```suggestion
```

Callable once `block.timestamp >= maturity`, stores the current `exchangeRate` as the `maturityRate` and "matures" the Principal Tokens. 

OPTIONAL - This method can be used to provide needed data for the calculation of interest post-maturity, but interfaces and other contracts MUST NOT expect this method to be present.

MUST emit the `Mature` event.

MUST revert if `block.timestamp < maturity`.

```yaml
- name: mature
  type: function
  stateMutability: nonpayable

  inputs: []

  outputs:
    - name: exchangeRate
      type: uint256
```

### Events

#### Mint

> `from` has exchanged `amount` of underlying for `amount` of Principal Tokens, and transferred those Principal Tokens to `to`.

Reviewer 2: If we remove `mint()`, this is unnecessary / would just be covered in the inherited ERCs?

Author: Yeah, it would be covered. Removed.

MUST be emitted when underlying is deposited and Principal Tokens are minted from the contract in the `ERC5070.mint` method.

```yaml
- name: Mint
  type: event

  inputs:
    - name: from
      indexed: true
      type: address
    - name: to
      indexed: true
      type: address
    - name: amount
      indexed: false
      type: uint256
```

#### Redeem

`from` has exchanged `amount` of Principal Tokens for `amount` of underlying, and transferred that underlying to `to`.

MUST be emitted when Principal Tokens are burnt and underlying is withdrawn from the contract in the `ERC5070.redeem` method.

```yaml
- name: Redeem
  type: event

  inputs:
    - name: from
      indexed: true
      type: address
    - name: to
      indexed: true
      type: address
    - name: amount
      indexed: false
      type: uint256
```

#### Mature

> _Alberto: Maybe we need this event, even if we don't have a `mature` function._

Reviewer 2: I think we do.

In many cases `maturity` doesnt actually mean the market has matured with reference to further yield generation and where it gets directed.

Author: Yeah, agreed.

> _Alberto: Maybe we need this event, even if we don't have a `mature` function._

Reviewer: ```suggestion
```

Author: Removed all of these.

The Principal Token has hit maturity, and a user has called `redeem` or `mature` on the Principal Token contract for the first time post-maturity.

OPTIONAL - This event can be used to improve usability and clarity for index based systems, but interfaces and other contracts MUST NOT expect this event to be present.

MUST be emitted when the `mature` method is called and the `maturityRate` is stored.

```yaml
- name: Mature
  type: event

  inputs:
    - name: timestamp
      indexed: true
      type: uint256
    - name: exchangeRate
      indexed: true
      type: uint256

> ```

Reviewer: I think it's customary to define these using solidity syntax, but that isn't a hard requirement.

Author: We don't want to discriminate against vyper or yul developers.

The optional `mint` method was also included to similarly cover the creation of Principal Tokens. Though a ubiquitous pattern for minting is less present in current implementations, and is less necessary than `redemption` for common use cases, the ability to directly mint Principal Tokens reduces the complexity of integrations.

The optional `maturityRate` and `exchangeRate` methods are included to allow for the accommodation of additional yield generation post-maturity. In implementations that include these optional methods, the amount of underlying returned by the `redeem` method is equivalent to ```amount + (amount * (exchangeRate/maturityRate))```.

## Backwards Compatibility

ERC-5070 is fully backward compatible with the ERC-20 standard and has no known compatibility issues with other standards.
For production implementations of Principal Tokens which do not use ERC-5070, wrapper adapters can be developed and used, or wrapped tokens can be implemented.

## Reference Implementations

> Official Example Implementation: https://github.com/JTraversa/Principal-Token-Standard

Reviewer: If the reference implementation is reasonably sized, you can include it directly in the `../assets/eip-5089` directory.

Author: REplaced with a TBA for now.

Yield fyTokens: https://github.com/yieldprotocol/vault-v2/blob/master/contracts/FYToken.sol

Swivel zcTokens: https://github.com/Swivel-Finance/swivel/blob/main/contracts/v2/marketplace/ZcToken.sol

> Element Principal Tokens: https://github.com/element-fi/elf-contracts/blob/main/contracts/Tranche.sol (maturity->unlockTimestamp)

Reviewer: Please remove all external links.

Author: Removed.

## Security Considerations

> Security considerations surrounding fees?

Reviewer: Is this a question for the reader or a note to self?

Author: Note to self, removed.

Fully permissionless use cases could fall prey to malicious implementations which only conform to an ERC-5070 interface but not the specification, failing to implement proper custodial functionality but offering the ability to purchase Principal Tokens through secondary markets.

It is recommended that all integrators review each implementation for potential ways of losing user deposits before integrating.

The methods `totalAssets`, `convertToShares` and `convertToAssets` are estimates useful for display purposes,

> and do _not_ have to confer the _exact_ amount of underlying assets their context suggests. _Alberto: This might be a remnant of ERC4626, which might be useful to explore. Maybe we use the same pattern instead of `exchangeRate`._

Reviewer: ```suggestion
and do _not_ have to confer the _exact_ amount of underlying assets their context suggests.
```
